commit 06204f0c9f539fcb8cb532166656e80b81bd689a
Author: Antonio Ospite <ao2@ao2.it>
Date:   Tue Oct 16 14:09:52 2018 +0200

    eval: make traps work when "set -e" is enabled
    
    When "set -e" is enabled traps are not always executed, in particular
    the EXIT trap is not executed when the shell exits on an unhandled
    error.
    
    Consider the following test script:
    
      #!/bin/dash
    
      set -e
    
      trap 'ret=$?; echo "EXIT: $ret"' EXIT
      trap 'exit 2' HUP INT QUIT PIPE TERM
    
      read variable
    
    By pressing Ctrl-C one would expect the EXIT trap to be called, as it is
    the case with other shells (bash, zsh), but dash does not do it.
    
    By calling dotrap() before jumping to the exit path when checkexit is
    not zero, dash behaves like other shells.
    
    Signed-off-by: Antonio Ospite <ao2@ao2.it>
    Signed-off-by: Herbert Xu <herbert@gondor.apana.org.au>

diff --git a/src/eval.c b/src/eval.c
index 546ee1b..dde9fa2 100644
--- a/src/eval.c
+++ b/src/eval.c
@@ -307,11 +307,11 @@ setstatus:
 		break;
 	}
 out:
+	dotrap();
+
 	if (checkexit & status)
 		goto exexit;
 
-	dotrap();
-
 	if (flags & EV_EXIT) {
 exexit:
 		exraise(EXEXIT);
